be/src/exec/pipeline/pipeline_driver_queue.cpp

```diff
@@ -74,8 +88,218 @@ StatusOr<DriverRawPtr> QuerySharedDriverQueue::take(size_t* queue_index) {
+    int queue_idx = -1;
+    double target_accu_time = 0;
+    DriverRawPtr driver_ptr;
+
+    for (int i = 0; i < QUEUE_SIZE; ++i) {
+        // we just search for queue has element
+        if (!_queues[i].queue.empty()) {
+            double local_target_time = _queues[i].accu_time_after_divisor();
+            // if this is first queue that has element, we select it;
+            // else we choose queue that the execution time is less sufficient,
+            // and record time.
+            if (queue_idx < 0 || local_target_time < target_accu_time) {
+                target_accu_time = local_target_time;
+                queue_idx = i;
+            }
+        }
+    }
+
+    // Always return non-null driver, which is guaranteed be the callee, e.g. DriverQueueWithWorkGroup::take().
+    DCHECK(queue_idx >= 0);
+
+    driver_ptr = _queues[queue_idx].queue.front();
+    _queues[queue_idx].queue.pop();
+
+    --_size;
+
+    return driver_ptr;
+}
+
+void QuerySharedDriverQueueWithoutLock::update_statistics(const DriverRawPtr driver) {
+    _queues[driver->get_dispatch_queue_index()].update_accu_time(driver);
+}
+
+void QuerySharedDriverQueueWithoutLock::_put_back(const DriverRawPtr driver) {
+    int level = driver->driver_acct().get_level() % QUEUE_SIZE;
+    driver->set_dispatch_queue_index(level);
+    _queues[level].queue.emplace(driver);
+    ++_size;
+}
+
+void DriverQueueWithWorkGroup::close() {
+    std::lock_guard<std::mutex> lock(_global_mutex);
+    _is_closed = true;
+    _cv.notify_all();
+}
+
+void DriverQueueWithWorkGroup::put_back(const DriverRawPtr driver) {
+    std::lock_guard<std::mutex> lock(_global_mutex);
+    _put_back<false>(driver);
+}
+
+void DriverQueueWithWorkGroup::put_back(const std::vector<DriverRawPtr>& drivers) {
+    std::lock_guard<std::mutex> lock(_global_mutex);
+
+    for (const auto driver : drivers) {
+        _put_back<false>(driver);
+    }
+}
+
+void DriverQueueWithWorkGroup::put_back_from_dispatcher(const DriverRawPtr driver) {
+    std::lock_guard<std::mutex> lock(_global_mutex);
+    _put_back<true>(driver);
+}
+
+void DriverQueueWithWorkGroup::put_back_from_dispatcher(const std::vector<DriverRawPtr>& drivers) {
+    std::lock_guard<std::mutex> lock(_global_mutex);
+
+    for (const auto driver : drivers) {
+        _put_back<true>(driver);
+    }
+}
+
+StatusOr<DriverRawPtr> DriverQueueWithWorkGroup::take(int dispatcher_id) {
+    std::unique_lock<std::mutex> lock(_global_mutex);
+
+    if (_is_closed) {
+        return Status::Cancelled("Shutdown");
+    }
+
+    while (_ready_wgs.empty()) {
+        _cv.wait(lock);
+        if (_is_closed) {
+            return Status::Cancelled("Shutdown");
+        }
+    }
+
+    // Try to take driver from any owner workgroup first.
+    auto wg = _find_min_owner_wg(dispatcher_id);
+    if (wg == nullptr) {
+        // All the owner workgroups don't have ready drivers, so select the other workgroup.
+        wg = _find_min_wg();
```

Comment:
Could the returned wg is nullptr?

Reply:
@mofeiatwork  No, if the code arrives here after breaking `while (_ready_wgs.empty()) {}`, `_ready_wgs` must contain at least wg.